src/Features/Core/Portable/CustomMessageHandler/CustomMessageHandlerService.cs

```diff
@@ -0,0 +1,182 @@
+    [Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
+    public CustomMessageHandlerService(ICustomMessageHandlerFactory customMessageHandlerFactory)
+    {
+        _customMessageHandlerFactory = customMessageHandlerFactory;
+    }
+
+    public async ValueTask<string> HandleCustomMessageAsync(
+        Solution solution,
+        string assemblyFolderPath,
+        string assemblyFileName,
+        string typeFullName,
+        string jsonMessage,
+        DocumentId? documentId,
+        CancellationToken cancellationToken)
+    {
+#if DEBUG
+        System.Diagnostics.Debugger.Launch();
+#endif
+
+        _ = assemblyFolderPath ?? throw new ArgumentNullException(nameof(assemblyFolderPath));
+        _ = assemblyFileName ?? throw new ArgumentNullException(nameof(assemblyFileName));
+        _ = typeFullName ?? throw new ArgumentNullException(nameof(typeFullName));
+        _ = jsonMessage ?? throw new ArgumentNullException(nameof(jsonMessage));
+
+        var assemblyPath = Path.Combine(assemblyFolderPath, assemblyFileName);
+
+        Document? document = null;
+        if (documentId is not null)
+        {
+            document = solution.GetDocument(documentId) ?? await solution.GetSourceGeneratedDocumentAsync(documentId, cancellationToken).ConfigureAwait(false);
+            Contract.ThrowIfNull(document);
+        }
+
+        try
+        {
+            ICustomMessageHandlerWrapper? handler;
+            lock (_lockObject)
+            {
+                // Check if the assembly is already loaded.
+                if (!_extensions.TryGetValue(assemblyFolderPath, out var extension))
+                {
+                    var loadContext = new System.Runtime.Loader.AssemblyLoadContext(name: $"RemoteCustomMessageHandlerService assembly load context for {assemblyFolderPath}", isCollectible: true);
+                    loadContext.Resolving += ResolveExtensionAssembly;
+
+                    extension = new CustomMessageHandlerExtension(loadContext);
+                    _extensions[assemblyFolderPath] = extension;
+                }
+
+                if (!extension.Handlers.TryGetValue(typeFullName, out handler))
+                {
+                    var assembly = extension.AssemblyLoadContext.LoadFromAssemblyPath(assemblyPath);
+                    var type = assembly.GetType(typeFullName)
+                        ?? throw new InvalidOperationException($"Cannot find type {typeFullName} in {assemblyPath}.");
+
+                    handler = _customMessageHandlerFactory.Create(type);
+                }
+            }
+
+            var message = JsonSerializer.Deserialize(jsonMessage, handler.MessageType);
+
+            var result = await handler.ExecuteAsync(message, document, solution, cancellationToken)
+                .ConfigureAwait(false);
+
+            var responseJson = JsonSerializer.Serialize(result, handler.ResponseType);
+
+            return responseJson;
+        }
+        catch (Exception e)
```

Comment:
Use exception filter to log the error instead of re-throwing. Rethrow causes the stack to be unwound.

Reply:
The desired behavior is for an exception in the plugin to be surfaced all the way up to the VS extension. The Roslyn plugin is code written by the extender and invoked by the extension, so it is reasonable that, if the plugin throws an exception, the exception is surfaced all the way to the caller in the VS extension.
I'm interested in your feedback about this.

Reply:
Absolutely. I'm not saying we should swallow it, quite opposite. We should use filter to *preserve* the all stack frames, so that one can inspect them in a crash dump.

Like so:

```
try
{ ... }
catch (Exception e) when (LogAndPropagate(e))
{
  // unreachable
}

bool LogAndPropagate(Exception e)
{
  Logger.Log(...);
  return false;
}
```

Reply:
I did this for the `UnloadCustomMessageHandlersAsync` method below.
This code was refactored and now the `catch` block has side effects beside logging
```
extension.Handlers[typeFullName] = null;
```
It seems strange to have such side effects hidden by a `when` clause. So I kept the `throw;` in this case. Let me know if you would still go for the `when` clause here.